v0.37 changelog - noise models (#5936)
PennyLaneAI/pennylane: doc/releases/changelog-0.37.0.md

<h4>Add noise models to your quantum circuits 📺</h4>

> * A new `qml.noise` module has been added, which contains utilities for building `NoiseModel`s 
>   and an `add_noise` transform for adding it to quantum circuits.

Author: ```suggestion
* Support for building noise models and applying them to a quantum circuit has been added
via the `NoiseModel` class and an `add_noise` transform.
```

I opt for just removing the last sentence then. sounds like we're shoehorning that bit in!

  * `qml.noise.op_in(ops)`: if any operators in `ops` are encountered in the circuit, add noise.
  * `qml.noise.wires_eq(wires)`: if an operator is applied to `wires`, add noise.
  * `qml.noise.wires_in(wires)`: if an operator is applied to any wire in `wires`, add noise.
  * custom noise conditions: custom conditions can be defined as functions decorated with `qml.BooleanFn` that return a Boolean value. For example, the following function will insert noise if a `qml.RY` operator is encountered with an angle of rotation that is less than `0.5`:

    ```python
    @qml.BooleanFn
    def c0(op):
        return isinstance(op, qml.RY) and op.parameters[0] < 0.5
    ```

  `op_eq`, `op_in`, `wires_eq`, and `wires_in` can also be combined together with `&`, `and`, `|`, etc.
  Once the conditions under which noise is to be inserted have been stated, we can specify exactly what 
  noise is inserted with the following:

  * `qml.noise.partial_wires(op)`: insert `op` on the wires that are specified by the condition that triggers adding this noise
  * custom noise operations: custom noise can be specified by defining a standard quantum function like below.
  

>   ```python
>   def n0(op, **kwargs):
>       qml.RY(op.parameters[0] * 0.05, wires=op.wires)
>   ```

Reviewer: ```suggestion
    ```python
    def n0(op, **kwargs):
        qml.RY(op.parameters[0] * 0.05, wires=op.wires)
    ```
```

  With that, we can create a `qml.NoiseModel` object whose argument must be a dictionary mapping conditions
  to noise:

  ```python

>   c1 = qml.noise.op_eq(qml.X) | qml.noise.wires_in([0, 1])

Reviewer: ```suggestion
  c1 = qml.noise.op_eq(qml.X) & qml.noise.wires_in([0, 1])
```
Needed this else the example fails because there is an `AmplitudeDamping` that has more than one wire passed.

  n1 = qml.noise.partial_wires(qml.AmplitudeDamping, 0.4)

  noise_model = qml.NoiseModel({c0: n0, c1: n1})
  ```

  ```pycon
  >>> noise_model
  NoiseModel({
      BooleanFn(c0): n0
      OpEq(PauliX) | WiresIn([0, 1]): AmplitudeDamping(gamma=0.4)
  })
  ```

  The noise model created can then be added to a QNode with `qml.add_noise`:

  ```python
  dev = qml.device("lightning.qubit", wires=3)

  @qml.qnode(dev)
  def circuit():
      qml.Y(0)
      qml.CNOT([0, 1])
      qml.RY(0.3, wires=2) # triggers c0
      qml.X(1) # triggers c1
      return qml.state()
  ```

  ```pycon
  >>> print(qml.draw(circuit)())
  0: ──Y────────╭●────┤  State
  1: ───────────╰X──X─┤  State
  2: ──RY(0.30)───────┤  State
  >>> circuit = qml.add_noise(circuit, noise_model)
  >>> print(qml.draw(circuit)())
  0: ──Y────────╭●───────────────────────────────────┤  State
  1: ───────────╰X─────────X──AmplitudeDamping(0.40)─┤  State

>   2: ──RY(0.30)──RY(0.01)────────────────────────────┤  State

Reviewer: Nice!

  ```

>   To learn more about this new functionality, check out our [`noise` module documentation](https://docs.pennylane.ai/en/stable/code/qml_noise.html)

Reviewer: ```suggestion
  To learn more about this new functionality, check out our [noise module documentation](https://docs.pennylane.ai/en/stable/code/qml_noise.html)
```

>   and keep your eyes peeled for an in-depth demo!

Reviewer: What do you think about mentioning the `level` argument :thinking:
https://xanaduai-pennylane--5936.com.readthedocs.build/en/5936/code/api/pennylane.add_noise.html

>   To learn more about this new functionality, check out our [noise module documentation](https://docs.pennylane.ai/en/stable/code/qml_noise.html)
>   and keep your eyes peeled for an in-depth demo!

Author: I suspect that this might be a little too much detail to get into here, which is why I linked to the docs. What about just a couple sentences:

```suggestion
  If more than one transform is applied to a QNode, control over when/where the `add_noise` transform is applied 
  in relation to the other transforms can be specified with the `level` keyword argument. By default, `add_noise` is applied
  after all the transforms that have been manually applied to the QNode until that point.
  To learn more about this new functionality, check out our [noise module documentation](https://docs.pennylane.ai/en/stable/code/qml_noise.html)
  and keep your eyes peeled for an in-depth demo!
```

Author: @trbromley

Reviewer: Thanks, it's great! :+1:
